lib/services/unified_resource.go

```diff
@@ -111,72 +121,163 @@ func (cfg *UnifiedResourceCacheConfig) CheckAndSetDefaults() error {
+func (c *UnifiedResourceCache) delete(ctx context.Context, res types.Resource) error {
+	key := resourceKey(res)
+
+	// delete generally only sends the id, so we will fetch the actual resource from our resources
+	// map and generate our sort keys. Then we can delete from the map and all the trees at once
+	resource := c.resources[key]
+
+	sortKey := makeResourceSortKey(resource)
+
+	return c.read(ctx, func(cache *UnifiedResourceCache) error {
+		if _, ok := cache.nameTree.Delete(&item{Key: sortKey.ByName}); !ok {
+			return trace.NotFound("key %q is not found in unified cache name sort tree", string(sortKey.ByName))
+		}
+		if _, ok := cache.typeTree.Delete(&item{Key: sortKey.ByType}); !ok {
+			return trace.NotFound("key %q is not found in unified cache type sort tree", string(sortKey.ByType))
 		}
+		// delete from resource map
+		delete(c.resources, key)
 		return nil
 	})
 }
 
-func (c *UnifiedResourceCache) getRange(ctx context.Context, startKey, endKey []byte, limit int) ([]resource, error) {
-	if len(startKey) == 0 {
-		return nil, trace.BadParameter("missing parameter startKey")
+func (c *UnifiedResourceCache) getSortTree(sortField string) (*btree.BTreeG[*item], error) {
+	switch sortField {
+	case SortByName:
+		return c.nameTree, nil
+	case SortByKind:
+		return c.typeTree, nil
+	default:
+		return nil, trace.NotImplemented("sorting by %v is not supporting in unified resources", sortField)
 	}
-	if len(endKey) == 0 {
-		return nil, trace.BadParameter("missing parameter endKey")
+
+}
+
+func (c *UnifiedResourceCache) getRange(ctx context.Context, startKey []byte, matchFn func(types.ResourceWithLabels) (bool, error), req *proto.ListUnifiedResourcesRequest) ([]resource, string, error) {
+	if len(startKey) == 0 {
+		return nil, "", trace.BadParameter("missing parameter startKey")
 	}
-	if limit <= 0 {
-		limit = backend.DefaultRangeLimit
+	if req.Limit <= 0 {
+		req.Limit = backend.DefaultRangeLimit
 	}
 
 	var res []resource
-	err := c.read(ctx, func(tree *btree.BTreeG[*item]) error {
-		tree.AscendRange(&item{Key: startKey}, &item{Key: endKey}, func(item *item) bool {
-			res = append(res, item.Value)
-			if limit > 0 && len(res) >= limit {
+	var nextKey string
+	err := c.read(ctx, func(cache *UnifiedResourceCache) error {
+		tree, err := cache.getSortTree(req.SortBy.Field)
+		if err != nil {
+			return trace.Wrap(err, "getting sort tree")
+		}
+		var iterateRange func(lessOrEqual, greaterThan *item, iterator btree.ItemIteratorG[*item])
+		var endKey []byte
+		if req.SortBy.IsDesc {
+			iterateRange = tree.DescendRange
+			endKey = backend.Key(prefix)
+		} else {
+			iterateRange = tree.AscendRange
+			endKey = backend.RangeEnd(backend.Key(prefix))
+		}
+		iterateRange(&item{Key: startKey}, &item{Key: endKey}, func(item *item) bool {
+			// get resource from resource map
+			resourceFromMap := cache.resources[item.Value]
+			if resourceFromMap == nil {
+				// continue the ascend. maybe we should throw an error?
```

Comment:
If `nil` entries are not valid you could instead check if the resource exists in the map and if not return and keep ascending? And if the resource is nil then return an error